src/include/btree_inline.h

```diff
@@ -2336,36 +2338,54 @@ __wt_btcur_skip_page(
      */
     if (previous_state == WT_REF_DELETED && __wt_page_del_visible(session, ref->page_del, true)) {
         *skipp = true;
+        WT_STAT_CONN_DATA_INCR(session, cursor_tree_walk_del_page_skip);
         goto unlock;
     }
 
-    /*
-     * Look at the disk address, if it exists, and if the page is unmodified. We must skip this test
-     * if the page has been modified since it was reconciled, since neither the delete information
-     * nor the timestamp information is necessarily up to date.
-     */
-    if ((previous_state == WT_REF_DISK ||
-          (previous_state == WT_REF_MEM && !__wt_page_is_modified(ref->page))) &&
-      __wt_ref_addr_copy(session, ref, &addr)) {
+    if (previous_state == WT_REF_MEM && !__wt_page_is_modified(ref->page)) {
+        /*
+         * Lock the page to prevent concurrent checkpoint reconciliation on the same page while we
+         * verify whether the page may be skipped or not.
+         */
+        if (ref->page->modify != NULL) {
+            if (WT_PAGE_TRYLOCK(session, ref->page) != 0)
+                goto unlock;
+            page_locked = true;
+        }
+
+        clean_page = true;
+    }
+
+    /* Look at the disk address, if it exists. */
+    if ((previous_state == WT_REF_DISK || clean_page) && __wt_ref_addr_copy(session, ref, &addr)) {
         /* If there's delete information in the disk address, we can use it. */
         if (addr.del_set && __wt_page_del_visible(session, &addr.del, true)) {
             *skipp = true;
+            WT_STAT_CONN_DATA_INCR(session, cursor_tree_walk_del_page_skip);
             goto unlock;
         }
 
         /*
          * Otherwise, check the timestamp information. We base this decision on the aggregate stop
          * point added to the page during the last reconciliation.
          */
-        if (addr.ta.newest_stop_txn != WT_TXN_MAX && addr.ta.newest_stop_ts != WT_TS_MAX &&
+        if (WT_TIME_AGGREGATE_HAS_STOP(&addr.ta) &&
           __wt_txn_snap_min_visible(session, addr.ta.newest_stop_txn, addr.ta.newest_stop_ts,
             addr.ta.newest_stop_durable_ts)) {
             *skipp = true;
-            goto unlock;
+            WT_STAT_CONN_DATA_INCR(session, cursor_tree_walk_ondisk_del_page_skip);
```

Comment:
The last time I added the code, we deliberately went against adding these statistics as we expected them to be in the performance path. Do we need this granularity of the stats?

Reply:
Yes, I agree with you that these statistics can add performance overhead. These statistics can be useful to understand how frequently the skip logic is working or not. I feel that in the event of any problem with the skip logic, these statistics may be helpful in understanding the issue. I am fine with removing them also.

Reply:
Cool, let's see what other reviewers (Alex ?) say.